libraries/ESP8266WiFi/src/BearSSLHelpers.cpp

```diff
@@ -0,0 +1,830 @@
+        free(pk->key.rsa.n);
+        free(pk->key.rsa.e);
+        free(pk);
+        return nullptr;
+      }
+      memcpy(pk->key.rsa.n, rk->n, rk->nlen);
+      pk->key.rsa.nlen = rk->nlen;
+      memcpy(pk->key.rsa.e, rk->e, rk->elen);
+      pk->key.rsa.elen = rk->elen;
+      return pk;
+
+    case BR_KEYTYPE_EC:
+      // TODO - not parsed yet in .T0 file
+      return nullptr;
+
+    default:
+      return nullptr;
+  }
+}
+
+void free_public_key(public_key *pk)
+{
+  if (pk) {
+    if (pk->key_type == BR_KEYTYPE_RSA) {
+      free(pk->key.rsa.n);
+      free(pk->key.rsa.e);
+    } else if (pk->key_type == BR_KEYTYPE_EC) {
+      // TODO - EC public keys not implemented
+    }
+    free(pk);
+  }
+}
+
+static private_key *decode_private_key(const unsigned char *buff, size_t len)
+{
+  std::unique_ptr<br_skey_decoder_context> dc(new br_skey_decoder_context); // auto-delete on exit
+  if (!dc.get()) return nullptr;
+
+  private_key *sk = nullptr;
+
+  br_skey_decoder_init(dc.get());
+  br_skey_decoder_push(dc.get(), buff, len);
+  int err = br_skey_decoder_last_error(dc.get());
+  if (err != 0) {
+    return nullptr;
+  }
+
+  const br_rsa_private_key *rk = nullptr;
+  const br_ec_private_key *ek = nullptr;
+  switch (br_skey_decoder_key_type(dc.get())) {
+    case BR_KEYTYPE_RSA:
+      rk = br_skey_decoder_get_rsa(dc.get());
+      sk = (private_key*)malloc(sizeof *sk);
+      if (!sk) {
+        return nullptr;
+      }
+      sk->key_type = BR_KEYTYPE_RSA;
+      sk->key.rsa.p = (uint8_t*)malloc(rk->plen);
+      sk->key.rsa.q = (uint8_t*)malloc(rk->qlen);
+      sk->key.rsa.dp = (uint8_t*)malloc(rk->dplen);
+      sk->key.rsa.dq = (uint8_t*)malloc(rk->dqlen);
+      sk->key.rsa.iq = (uint8_t*)malloc(rk->iqlen);
+      if (!sk->key.rsa.p || !sk->key.rsa.q || !sk->key.rsa.dp || !sk->key.rsa.dq || !sk->key.rsa.iq) {
+        free_private_key(sk);
+        return nullptr;
+      }
+      sk->key.rsa.n_bitlen = rk->n_bitlen;
+      memcpy(sk->key.rsa.p, rk->p, rk->plen);
+      sk->key.rsa.plen = rk->plen;
+      memcpy(sk->key.rsa.q, rk->q, rk->qlen);
+      sk->key.rsa.qlen = rk->qlen;
+      memcpy(sk->key.rsa.dp, rk->dp, rk->dplen);
+      sk->key.rsa.dplen = rk->dplen;
+      memcpy(sk->key.rsa.dq, rk->dq, rk->dqlen);
+      sk->key.rsa.dqlen = rk->dqlen;
+      memcpy(sk->key.rsa.iq, rk->iq, rk->iqlen);
+      sk->key.rsa.iqlen = rk->iqlen;
+      return sk;
+
+    case BR_KEYTYPE_EC:
+      ek = br_skey_decoder_get_ec(dc.get());
+      sk = (private_key*)malloc(sizeof *sk);
+      sk->key_type = BR_KEYTYPE_EC;
+      sk->key.ec.curve = ek->curve;
+      sk->key.ec.x = (uint8_t*)malloc(ek->xlen);
+      if (!sk->key.ec.x) {
+        free_private_key(sk);
+        return nullptr;
+      }
+      memcpy(sk->key.ec.x, ek->x, ek->xlen);
+      sk->key.ec.xlen = ek->xlen;
+      return sk;
+
+    default:
+      return nullptr;
+  }
+}
+
+void free_private_key(private_key *sk)
+{
+  if (sk) {
+    switch (sk->key_type) {
```

Comment:
default case (error handling)

Reply:
While I could have a default case, I'd have nothing to do here as the switch is just figuring out which part of the union to use to go and free elements.  If it's not one of EC or RSA I don't have anything to free (could be uninitted, etc.)

Reply:
The goal of the default case in code like this is error handling, or at least an assert, i.e.: the default body should never execute. If it ever does execute, you'll know that something went terribly wrong (i.e.: mem corruption, uninit, etc)